routing+routerrpc: improve prob. estimation for untried connections

joostjager opened the pull request:
This PR changes mission control to partially base the estimated probability for untried connections on historical results obtained in previous payment attempts. This prevents the payment loop from getting stuck in an area of the network that is not of good quality. It also incentivizes routing nodes to keep all
of their channels in good shape.

The configuration parameter `--routerrpc.aprioriweight` is a value in the range [0, 1] that defines to what extent historical results should be extrapolated to untried connections. Setting it to 1 will completely ignore historical results and always assume the configured a priori probability for untried connections. A value of 0 will ignore the a priori probability completely and only base the probability on historical results, unless there are none available.

Using this parameter, senders can configure how aggressive nodes should be penalized. A low a priori weight will make mission control switch quickly to using a different node if a tried node doesn't perform well. The downside of this is that the sender may miss out on low fee channels that could possibly complete the payment successfully.

Indirectly, the collective of senders also shapes the network with their settings of this parameter. When senders start to demand a higher performance from routing nodes, routing nodes need to meet that expectation in order to keep generating revenue.

joostjager commented:
Did a bit of performance testing with and without these changes. With a clean mission control history, there is no noticeable difference. This is expected, because there are hardly iterations in the loop inside `GetNodeProbability`. For the test call to `QueryRoutes`, 330 nodes were visited and 7661 edges expanded.

To see a performance effect, it is probably needed to have a fully loaded mission control history. It is a question how likely this situation is, especially on low-powered devices where pathfinding performance becomes important. There is also a default limit to the mission control history of 1000 past payments, although this can be changed.

I tried a simulation of a full mission control by inserting 10 fake results for every node visited, but didn't see an effect either. Db access during path finding is likely still where time is mainly spent.

I'd propose to not prematurely add optimizations like caching of the node probability in this pr and first see what feedback we get on it from users. I've added another commit that logs useful performance metrics.

halseth commented on routing/missioncontrol.go:
```diff
@@ -246,22 +254,51 @@ func (m *MissionControl) GetProbability(fromNode, toNode route.Vertex,
 
-	return probability
+	now := m.now()
+	for _, result := range results {
```
In particular I'm interested in the scenarios we talked about earlier:
A node having 100s of untried channels, and 1 tried channel (success/failure)
A node having 100s of untried channels, and X tried channels (success/failure), where X is about 2-5
A node having few channels, and some result (say 3 channels, 2 failures/successes).

joostjager replied:
Extracted probability estimation from mission control and added more unit tests.

The number of untried channels isn't a factor in the calculation at all, so I can't write a unit test for that. Probability estimates are only based on previous payment results.

joostjager commented on routing/missioncontrol.go:
```diff
@@ -246,22 +254,51 @@ func (m *MissionControl) GetProbability(fromNode, toNode route.Vertex,
+		totalWeight += weight
+	}
+
+	return total / totalWeight
```
Note to self: handle `totalWeight == 0` case

joostjager replied:
Note to self: explain single observation case

joostjager commented on routing/missioncontrol.go:
```diff
@@ -60,7 +60,7 @@ const (
 type MissionControl struct {
 	// lastPairResult tracks the last payment result per node pair.
-	lastPairResult map[DirectedNodePair]timedPairResult
+	lastPairResult map[route.Vertex]map[route.Vertex]timedPairResult
```
commit message added

joostjager commented on routing/missioncontrol.go:
```diff
@@ -254,16 +254,25 @@ func (m *MissionControl) getProbAfterFail(lastFailure time.Time) float64 {
 
+// getWeight calculates a weight in the range [0, 1] that should be assigned to
+// a payment result.Weight follows an exponential curve that starts at 1 when
+// the result is freshed and asymptotically approaches zero over time. The rate
```
yes, you know, freshing. fixed :smile:

joostjager commented on routing/missioncontrol.go:
```diff
@@ -246,22 +254,51 @@ func (m *MissionControl) GetProbability(fromNode, toNode route.Vertex,
+	// The value of the apriori weight is in the range [0, 1]. Convert it to
+	// a factor that properly expresses the intention of the weight in the
+	// following weight average calculation.
+	aprioriFactor := 1/(1-m.cfg.AprioriWeight) - 1
```
explanation added

joostjager commented on routing/missioncontrol.go:
```diff
@@ -246,22 +254,51 @@ func (m *MissionControl) GetProbability(fromNode, toNode route.Vertex,
+	// to make sure that all (not just some) of their channels are in good
+	// shape. Senders will steer around nodes that have shown a few
+	// failures, even though there may be many channels still untried.
+	total := m.cfg.AprioriHopProbability * aprioriFactor
```
renamed to `probabilitiesTotal`

joostjager commented on routing/missioncontrol.go:
```diff
@@ -303,40 +340,62 @@ func (m *MissionControl) setLastPairResult(fromNode,
 
+// setAllPairResult stores a result for all known connection of the given node.
+func (m *MissionControl) setAllPairResult(fromNode route.Vertex,
+	result *timedPairResult) {
```
This is how we do it for the other functions too. Do you think it will be faster as a value?

halseth replied:
Was just so you dont have to dereference below.

joostjager replied:
Fixed

Roasbeef commented on routing/result_interpretation.go:
```diff
@@ -367,7 +367,19 @@ func (i *interpretedResult) processPaymentOutcomeUnknown(route *route.Route) {
 // failNode marks the node indicated by idx in the route as failed. This
 // function intentionally panics when the self node is failed.
 func (i *interpretedResult) failNode(rt *route.Route, idx int) {
+	// Mark the node as failing.
```
`idx-1` can be lifted to a new local variable, say `targetNode` perhaps.

joostjager replied:
Technically it could, both `idx-1` expression don't have the exact same meaning. It is an index into the route hops list and a route channel index.

Roasbeef commented on routing/result_interpretation.go:
```diff
@@ -367,7 +367,19 @@ func (i *interpretedResult) processPaymentOutcomeUnknown(route *route.Route) {
+	i.pairResults[inPair.Reverse()] = pairResult{}
+
+	// If not the ultimate node, mark the outgoing connection as failed for
+	// the node.
```
If we have `A <-> B <-> C` and `B` failed, the incoming is ` A -> B` and outgoing is `B -> C`?

joostjager replied:
Yes. Added a helper variables to make this clearer.

Roasbeef commented on routing/missioncontrol.go:
```diff
@@ -377,28 +333,21 @@ func (m *MissionControl) GetHistorySnapshot() *MissionControlSnapshot {
-			OtherSuccessProb: otherProb,
-		})
-	}
+	nodes := make([]MissionControlNodeSnapshot, 0)
```
Style nit: can instead just declare it (guessing you hit the "should pre allocate" directive in the liter):
```go
var nodes []MissionControlSnapshot
```

Roasbeef replied:
If we _really_ want to pre-allocate though, we can add a method or function to compute the total size of the multi-dimensional map.

joostjager replied:
Converted to var initializer. No, I don't necessarily want to pre-allocate :smile:

Roasbeef commented on routing/probability_estimator.go:
```diff
@@ -0,0 +1,154 @@
+			// Weigh success with a constant high weight of 1. There
+			// is no decay.
+			totalWeight++
+			probabilitiesTotal += prevSuccessProbability
```
Should be `p.prevSuccessProbability`.

joostjager commented:
> Completed an initial surface-level pass. Will do another one focusing on the core estimation changes to ensure I fully grok them. Overall I think this is a step in the right direction, but wonder how we can test these assumptions in the wild, and also if it's too early to make a change like this given that as we know it, node operators aren't very sophisticated _yet_.

We can always set the default a priori weight quite high, so the effect of this will be light initially. Give some time for users and operators to become aware of the new scoring.

halseth commented on routing/missioncontrol.go:
```diff
@@ -505,11 +460,31 @@ func (m *MissionControl) applyPaymentResult(
-		m.lastNodeFailure[*i.nodeFailure] = result.timeReply
+		m.setAllPairResult(*i.nodeFailure, &timedPairResult{
+			timestamp:  result.timeReply,
+			pairResult: pairResult{},
```
same comment about setting `success=false`

halseth commented on routing/probability_estimator.go:
```diff
@@ -0,0 +1,154 @@
+	// If there is no history for this pair, return the node probability
+	// that is a probability estimate for untried channel.
+	if !ok {
+		return nodeProbability
```
maybe defer gettting `nodeProbsbility` to where it is actually returned, to avoid doing result iteration unnecessarily

halseth commented on routing/probability_estimator_test.go:
```diff
@@ -0,0 +1,162 @@
+func TestProbabilityEstimatorNoResults(t *testing.T) {
+	ctx := newEstimatorTestContext(t)
+
+	ctx.assertPairProbability(testTime, 0, 0, 0.6)
```
use `aprioriHopProb`?

joostjager replied:
fixed

halseth commented on routing/probability_estimator_test.go:
```diff
@@ -0,0 +1,162 @@
+		},
+	}
+
+	// For an untried node, we expected the node probability. The weight for
```
Good explanations in these test cases 👍

halseth commented on routing/probability_estimator_test.go:
```diff
@@ -0,0 +1,162 @@
+	expectedNodeProb := (3*aprioriHopProb + 1*prevSuccessProbability) /
+		(3 + 1 + 0.25 + 0.125)
+
+	ctx.assertPairProbability(testTime, untriedNode, 100, expectedNodeProb)
```
it is a bit confusing that `untriedNode` and node `1` get different results here, since neither are referenced. I think the index in `ctx.result` is used to denote the `toNode` of the result. Could we instead set this explicitly (by making `results ` a map), and extract `node0 = 0`, `node1 = 1` into constants for readability?

joostjager replied:
Yes it is a bit obscure. Followed your suggestion

halseth commented on routing/pathfind.go:
```diff
@@ -296,6 +297,18 @@ func findPath(g *graphParams, r *RestrictParams, cfg *PathFindingConfig,
+	// Pathfinding can be a significant portion of the total payment
+	// latency, especially on low-powered devices. Log several metrics to
+	// aid in the analysis performance problems in this area.
+	start := time.Now()
```
nice, we should consider exposing these metrics in the future.

joostjager replied:
Where would you want to expose it in the future?

joostjager replied:
I am not sure if it is very informative, because it depends mainly on payment details.

halseth replied:
It depends on uniform payments being made regularly indeed, but I can imagine it being useful in some cases. But was just an idea for future work.

joostjager commented on routing/result_interpretation.go:
```diff
@@ -367,7 +367,19 @@ func (i *interpretedResult) processPaymentOutcomeUnknown(route *route.Route) {
+
+	// Mark the incoming connection as failed for the node.
+	inPair, _ := getPair(rt, idx-1)
+	i.pairResults[inPair.Reverse()] = pairResult{}
```
If I remember correctly, there were comments in other prs about not adding redundant initializers. Not sure what you standard is here.

halseth replied:
Personally I like adding them (exceptions being `nil` initializers perhaps), and in this case I feel it makes sense, since an empty struct is not obvious what it means.

joostjager commented on routing/missioncontrol.go:
```diff
@@ -60,7 +60,7 @@ const (
 type MissionControl struct {
 	// lastPairResult tracks the last payment result per node pair.
-	lastPairResult map[DirectedNodePair]timedPairResult
+	lastPairResult map[route.Vertex]map[route.Vertex]timedPairResult
```
Fixed

joostjager commented:
I realized that the calculation of the node probability should also be amount independent. Especially if we make the amount more important in the follow up pr #3493. Updated `GetNodeProbability`. I also removed those precalculated probabilities from the mc state dump and provided a new rpc to query a specific probability.

joostjager commented:
Move new `querymcprob` call to #3556

joostjager commented on routing/result_interpretation.go:
```diff
@@ -367,7 +367,21 @@ func (i *interpretedResult) processPaymentOutcomeUnknown(route *route.Route) {
 	i.nodeFailure = &rt.Hops[idx-1].PubKeyBytes
+
+	// Mark the incoming connection as failed for the node.
+	incomingChannelIdx := idx - 1
```
See https://github.com/lightningnetwork/lnd/pull/3462/commits/ae91eb79e3959a08f2a2c1699194142144208b68#r326447064

joostjager commented on lnrpc/routerrpc/router.proto:
```diff
@@ -373,8 +355,7 @@ message PairHistory {
 
-    /// Estimation of success probability for this pair.
-    float success_prob = 5 [json_name = "success_prob"];
+    reserved 5;
```
Discussed offline. We don't want probabilities for an implicit amount that are prone to misinterpretation.

joostjager replied:
Yes, a follow up pr adds this info again, but in a separate rpc call that allows the caller the specify an amount.

The "assume" part was indeed light. The only assumption is that the probability applies to any amount greater than or equal to `min_penalize_amt`. This could also have been clarified with a comment. 

For the previously existing `other_success_prob` it was less clear to what amount it actually applied.

But bottom line is that with the changes in this and follow up prs, those assumptions don't hold anymore.

joostjager commented on routing/result_interpretation.go:
```diff
@@ -367,7 +367,21 @@ func (i *interpretedResult) processPaymentOutcomeUnknown(route *route.Route) {
 // failNode marks the node indicated by idx in the route as failed. This
 // function intentionally panics when the self node is failed.
 func (i *interpretedResult) failNode(rt *route.Route, idx int) {
```
Sentence added

joostjager commented on routing/missioncontrol.go:
```diff
@@ -446,8 +452,7 @@ func (m *MissionControl) applyPaymentResult(
 		m.setAllPairResult(*i.nodeFailure, timedPairResult{
-			timestamp:  result.timeReply,
-			pairResult: pairResult{},
+			timestamp: result.timeReply,
```
I've added constructors to make it more obvious without adding redundant field initializers.

joostjager commented on lnrpc/routerrpc/router.proto:
```diff
@@ -348,17 +348,21 @@ message PairHistory {
+
+    reserved 3, 4, 5, 6;
+
+    PairData data = 7 [json_name="data"];
```
Moved to #3556

halseth commented on routing/result_interpretation.go:
```diff
@@ -26,6 +26,20 @@ type pairResult struct {
 }
 
+// newFailPairResult creates a new result struct for a failure.
+func newFailPairResult(minPenalizeAmt lnwire.MilliSatoshi) pairResult {
```
nit: maybe `new` is redundant? Could do `failPairResult(99)` and `successPairResult` only

joostjager replied:
renamed

halseth commented on routing/missioncontrol.go:
```diff
@@ -226,100 +245,37 @@ func (m *MissionControl) GetProbability(fromNode, toNode route.Vertex,
-		if lastPairResult.success {
-			return prevSuccessProbability
-		}
+// setAllPairResult stores a result for all known connection of the given node.
```
should be `setAllFail`, since only used for failure?

joostjager replied:
It describes what the function does

halseth replied:
I think an earlier version of the PR changed this to `setAllFail`, since we never use it to set successes?

joostjager replied:
I don't remember exactly, but I think in that version the function could also only set a failure.

halseth replied:
question is why we need one that can set success if that is never used

joostjager replied:
yes, yagni, fixed.

Roasbeef commented on routing/probability_estimator.go:
```diff
@@ -0,0 +1,161 @@
+	probabilitiesTotal := p.aprioriHopProbability * aprioriFactor
+	totalWeight := aprioriFactor
+
+	for _, result := range results {
```
Latest version of this loop is much easier to follow 👍

Roasbeef commented on routing/probability_estimator.go:
```diff
@@ -0,0 +1,161 @@
+
+	// Getter for the node probability that is used for untried channels.
+	getNodeProbability := func() float64 {
+		return p.getNodeProbability(now, results, amt)
```
I don't think this function closure really adds much, other than introducing 3 extra lines in this diff.

joostjager replied:
It was an attempt to deduplicate code. But was in doubt already myself about it. Unrolled.

Roasbeef commented on server.go:
```diff
@@ -660,6 +660,7 @@ func newServer(listenAddrs []net.Addr, chanDB *channeldb.DB,
 			AprioriHopProbability: routingConfig.AprioriHopProbability,
 			PenaltyHalfLife:       routingConfig.PenaltyHalfLife,
 			MaxMcHistory:          routingConfig.MaxMcHistory,
+			AprioriWeight:         routing.DefaultAprioriWeight,
```
We should add this to the routing config so we can experiment with different a prior weight values without having to recompile the target `lnd` binary.

joostjager replied:
Definitely needs to be configurable. This is _the_ parameter that is going to shrink the graph, hopefully :) Fixed

joostjager commented on routing/missioncontrol.go:
```diff
@@ -60,7 +63,7 @@ const (
 type MissionControl struct {
 	// lastPairResult tracks the last payment result per node pair.
-	lastPairResult map[DirectedNodePair]timedPairResult
+	lastPairResult map[route.Vertex]NodeResults
```
excellent

joostjager commented on routing/result_interpretation.go:
```diff
@@ -378,10 +378,25 @@ func (i *interpretedResult) processPaymentOutcomeUnknown(route *route.Route) {
+	if idx < len(rt.Hops) {
+		outgoingChannelIdx := idx
+		outPair, _ := getPair(rt, outgoingChannelIdx)
+		i.pairResults[outPair] = failPairResult(0)
```
We don't need it for this PR. Here it is intentionally set to zero, because it should be a penalty to any amount, not just `>= amt`.

joostjager commented on routing/result_interpretation.go:
```diff
@@ -378,10 +378,25 @@ func (i *interpretedResult) processPaymentOutcomeUnknown(route *route.Route) {
+	// Mark the incoming connection as failed for the node.
+	incomingChannelIdx := idx - 1
+	inPair, _ := getPair(rt, incomingChannelIdx)
+	i.pairResults[inPair.Reverse()] = failPairResult(0)
```
We intent to penalize as much as we can for a node level failure, including future outgoing traffic for this connection. The pair as it is returned by getPair is directed towards the failed node. Therefore we first reverse the pair.

Added this as comment.

joostjager commented on routing/missioncontrol_test.go:
```diff
@@ -174,7 +179,7 @@ func TestMissionControl(t *testing.T) {
 
 	if len(history.Pairs) != 3 {
-		t.Fatalf("expected 2 pairs, but got %v", len(history.Pairs))
+		t.Fatalf("expected 3 pairs, but got %v", len(history.Pairs))
```
done

joostjager commented on routing/probability_estimator.go:
```diff
@@ -0,0 +1,156 @@
+// that have not been tried before. The results parameter is a list of last
+// payment results for that node.
+func (p *probabilityEstimator) getNodeProbability(now time.Time,
+	results map[route.Vertex]timedPairResult,
```
fixed

joostjager commented on routing/probability_estimator.go:
```diff
@@ -0,0 +1,156 @@
+// toNode based on historical payment outcomes for the from node. Those outcomes
+// are passed in via the results parameter.
+func (p *probabilityEstimator) getPairProbability(
+	now time.Time, results map[route.Vertex]timedPairResult,
```
fixed

joostjager commented on routing/probability_estimator_test.go:
```diff
@@ -0,0 +1,163 @@
+
+	c.t.Helper()
+
+	results := make(map[route.Vertex]timedPairResult)
```
fixed
